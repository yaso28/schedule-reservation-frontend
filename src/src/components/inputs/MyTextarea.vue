<template>
  <CTextarea
    :id="id"
    horizontal
    :label="label"
    :is-valid="validate"
    :invalid-feedback="invalidMessage"
    :value="value"
    :rows="rows"
    @input="setValue"
  />
</template>

<script>
export default {
  name: 'MyTextarea',
  mixins: [require('./input-mixin').default],
  props: {
    rows: {
      type: Number,
      default: 3
    }
  }
}
</script>
