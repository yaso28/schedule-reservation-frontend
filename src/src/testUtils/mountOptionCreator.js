import { createLocalVue } from '@vue/test-utils'
import VueRouter from 'vue-router'
import routes from '@/router/routes'
import Vuex from 'vuex'
import VueI18n from 'vue-i18n'
import i18nOptions from '@/i18n/options'
import VueMeta from 'vue-meta'
import CoreuiVue from '@coreui/vue'
import icons from '@/icons'
import services from '@/services'

class MountOptionCreator {
  create (storeModules = null) {
    const localVue = createLocalVue()

    // router
    localVue.use(VueRouter)
    const router = new VueRouter({
      mode: 'history',
      base: process.env.BASE_URL,
      routes
    })

    // i18n
    localVue.use(VueI18n)
    const i18n = new VueI18n(i18nOptions)

    // vue-meta
    localVue.use(VueMeta)

    // Core UI
    localVue.use(CoreuiVue)

    const option = {
      localVue,
      router,
      i18n,
      icons,
      provide: services
    }

    // store
    if (storeModules) {
      localVue.use(Vuex)
      option.store = new Vuex.Store({
        modules: storeModules
      })
    }

    return option
  }
}

export default new MountOptionCreator()
