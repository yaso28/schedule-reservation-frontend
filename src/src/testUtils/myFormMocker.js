import MyForm from '@/models/MyForm'
jest.mock('@/models/MyForm')

class MyFormMocker {
  createMockSingleField (field, initValue) {
    const input = {}
    input[field] = initValue
    return this.createMock(input)
  }

  createMock (input) {
    const mockForm = new MyForm(input)
    mockForm.values = input
    mockForm.errors = {}
    return mockForm
  }

  async setValue (wrapper, form, field, value) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        values: Object.assign({}, form.values, { [field]: value })
      })
    })
  }

  setValueNonProps (form, field, value) {
    form.values[field] = value
  }

  async setError (wrapper, form, field, errorMessage) {
    await wrapper.setProps({
      form: Object.assign({}, form, {
        errors: Object.assign({}, form.errors, { [field]: errorMessage })
      })
    })
  }
}

export default new MyFormMocker()
