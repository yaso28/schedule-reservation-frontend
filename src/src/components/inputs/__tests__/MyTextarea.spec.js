import { mount } from '@vue/test-utils'
import mountOptionCreator from '@/testUtils/mountOptionCreator'
import MyTextarea from '../MyTextarea'
import myFormMocker from '@/testUtils/myFormMocker'

jest.mock('@/models/MyForm')

function getInputWrapper (wrapper) {
  return wrapper.find('textarea')
}

describe('MyTextarea', () => {
  let form
  let field
  let wrapper
  beforeEach(() => {
    const label = 'サンプル'
    field = 'sample_field'
    form = myFormMocker.createMockSingleField(field, '')

    const mountOption = mountOptionCreator.create()
    mountOption.propsData = { label, form, field }
    wrapper = mount(MyTextarea, mountOption)
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  describe('通常', () => {
    test('表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('バリデーションエラー', () => {
    beforeEach(async () => {
      await myFormMocker.setError(wrapper, form, field, 'エラーです')
    })

    test('表示', () => {
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('入力', () => {
    const newValue = 'あいう\nえお'
    beforeEach(async () => {
      const inputWrapper = getInputWrapper(wrapper)
      await inputWrapper.setValue(newValue)
    })

    test('MyForm.setValue()を呼出', () => {
      expect(form.setValue).toHaveBeenCalledWith(field, newValue)
    })
  })

  describe('オプション', () => {
    describe('rows', () => {
      beforeEach(() => {
        wrapper.setProps({ rows: 10 })
      })

      test('rows属性変更', () => {
        expect(getInputWrapper(wrapper).attributes()).toMatchSnapshot()
      })
    })
  })
})
