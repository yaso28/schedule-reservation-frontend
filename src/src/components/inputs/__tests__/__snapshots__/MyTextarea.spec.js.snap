// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyTextarea オプション rows rows属性変更 1`] = `
Object {
  "class": "form-control",
  "id": "sample_field",
  "rows": "10",
}
`;

exports[`MyTextarea バリデーションエラー 表示 1`] = `
<div
  class="form-group form-row"
  role="group"
>
  <label
    class="col-form-label col-sm-3"
    for="sample_field"
  >
     サンプル 
  </label>
  <div
    class="col-sm-9"
  >
    <!---->
    <textarea
      class="form-control is-invalid"
      id="sample_field"
      rows="3"
    />
    <!---->
    <!---->
    <div
      class="invalid-feedback"
    >
      エラーです
    </div>
    <!---->
  </div>
</div>
`;

exports[`MyTextarea 通常 表示 1`] = `
<div
  class="form-group form-row"
  role="group"
>
  <label
    class="col-form-label col-sm-3"
    for="sample_field"
  >
     サンプル 
  </label>
  <div
    class="col-sm-9"
  >
    <!---->
    <textarea
      class="form-control"
      id="sample_field"
      rows="3"
    />
    <!---->
    <!---->
    <!---->
    <!---->
  </div>
</div>
`;
